[serve] autoscaling release tests

zcin opened the pull request:
[serve] autoscaling release tests

Add autoscaling load test as serve release test. Starts a service and runs a locust workload against it.
Name: `serve_autoscaling_load_test`.

Signed-off-by: Cindy Zhang <cindyzyx9@gmail.com>

aslonnie commented on release/ray_release/byod/byod_locust.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+# This script is used to build an extra layer on top of the base anyscale/ray image 
+
+pip install locust==2.16.1
```
why does this need to be installed separately? could you add a comment to explain?

and nit: maybe add an end of line?

zcin replied:
Resolved by https://github.com/ray-project/ray/pull/44392

edoakes commented on release/serve_tests/workloads/anyscale_service_utils.py:
```diff
@@ -0,0 +1,113 @@
+)
+
+
+def get_anyscale_cli_token() -> str:
+    try:
+        token, _ = AuthenticationBlock._load_credentials()
+        logger.info("Loaded anyscale credentials from local storage.")
+        return token
+    except Exception:
+        pass  # Ignore errors
+
+    logger.info("Missing ANYSCALE_CLI_TOKEN, retrieving from AWS secrets store")
+    return boto3.client("secretsmanager", region_name="us-west-2").get_secret_value(
+        SecretId=str(RELEASE_AWS_ANYSCALE_SECRET_ARN)
+    )["SecretString"]
```
is this the standard way to do this in the release tests? is there some shared util we can use from other release tests?

zcin replied:
Yeah this logic is duplicated in `ray_release/aws.py`, but that's not built into the ray wheel. @aslonnie what would be a good place to put this logic as a shared util?

zcin replied:
Update: I've removed this because it is unnecessary.

edoakes commented on release/serve_tests/workloads/anyscale_service_utils.py:
```diff
@@ -0,0 +1,113 @@
+
+def get_anyscale_cli_token() -> str:
+    try:
+        token, _ = AuthenticationBlock._load_credentials()
```
hmm we really shouldn't be reaching into private APIs like this

zcin replied:
Found that this was unnecessary, the credentials are already set in the job - removed.

edoakes commented on release/serve_tests/workloads/anyscale_service_utils.py:
```diff
@@ -0,0 +1,113 @@
+    return f"anyscale/image/{cluster_env.name}:{build.revision}"
+
+
+def check_service_state(sdk: AnyscaleSDK, service_id: str, expected_state: str):
+    state = sdk.get_service(service_id).result.current_state
+    print(f"Waiting for service {service_id} to be {expected_state}, currently {state}")
+    assert (
+        str(state) == expected_state
+    ), f"Service {service_id} is {state}, expected {expected_state}."
+    return True
```
let's use `anyscale.service.status` for this rather than the old SDK

zcin replied:
done!

edoakes commented on release/serve_tests/workloads/anyscale_service_utils.py:
```diff
@@ -0,0 +1,113 @@
+    )["SecretString"]
+
+
+def get_current_build_id(sdk: AnyscaleSDK) -> str:
+    cluster_id = os.environ["ANYSCALE_CLUSTER_ID"]
+    cluster = sdk.get_cluster(cluster_id)
+    return cluster.result.cluster_environment_build_id
+
+
+def build_id_to_image_uri(sdk: AnyscaleSDK, build_id: str) -> str:
+    build = sdk.get_cluster_environment_build(build_id).result
+    cluster_env = sdk.get_cluster_environment(build.cluster_environment_id).result
+    return f"anyscale/image/{cluster_env.name}:{build.revision}"
```
let's avoid using the old `SDK` at all (assume they're not public APIs) -- why do we need to get the build_id from current cluster at all

zcin replied:
A BYOD image is built for release test jobs that contains the ray compiled at that commit - I want to reuse that for the service cluster.

edoakes replied:
as discussed offline i think we should take this as an env variable

that way the test can be run from local as well

zcin replied:
Makes sense, I've addressed this. The cluster env name is now passed in as an env variable.

edoakes commented on release/serve_tests/workloads/autoscaling_load_test.py:
```diff
@@ -0,0 +1,127 @@
+
+@click.command()
+def main():
+    sdk = AnyscaleSDK()
```
again pls no old SDK

edoakes commented on release/serve_tests/workloads/autoscaling_load_test.py:
```diff
@@ -0,0 +1,127 @@
+        ray.init(address="auto")
+        status = service.status(name=service_name)
+
+        # Start the locust workload
+        num_locust_workers = int(ray.available_resources()["CPU"]) - 1
+        logger.info(f"Spawning {num_locust_workers} Locust worker Ray tasks.")
+        master_address = ray.util.get_node_ip_address()
+        worker_refs = []
+
+        # Start Locust workers
+        for _ in tqdm(range(num_locust_workers)):
+            locust_worker = LocustWorker.remote(
+                host_url=status.query_url,
+                token=status.query_auth_token,
+                master_address=master_address,
+                data={"uri": IMAGE_URI},
+            )
+            worker_refs.append(locust_worker.run.remote())
+
+        # Start Locust master
+        master_worker = LocustMaster.remote(
+            host_url=status.query_url,
+            token=status.query_auth_token,
+            expected_num_workers=num_locust_workers,
+            stages=locust_stages,
+        )
+        master_ref = master_worker.run.remote()
+
+        # Collect results and metrics
+        stats = ray.get(master_ref)
+        errors = sorted(chain(*ray.get(worker_refs)), key=lambda e: e["start_time"])
+
+        if stats.get("num_failures") > 0:
+            raise RuntimeError(
+                f"There were failed requests: {json.dumps(errors, indent=4)}"
+            )
+        else:
```
I think all of this stuff should get shoved into some kind of utility interface (maybe in `locust_utils.py`?). so each test will just do something like:
```python
results = run_locust_load_test(config: LocustLoadTestConfig)
```
where the load test config contains number of workers, data, query token, etc. etc.

zcin replied:
Good idea, I've hidden all of it into a function called `run_locust_load_test`.

edoakes commented on release/serve_tests/workloads/locust_utils.py:
```diff
@@ -0,0 +1,172 @@
+
+        stats_entry_key = ("", "GET")
+        stats_entry = master_runner.stats.entries.get(stats_entry_key)
+        return {
+            "history": master_runner.stats.history,
+            "total_requests": master_runner.stats.num_requests,
+            "num_failures": master_runner.stats.num_failures,
+            "avg_latency": stats_entry.avg_response_time,
+            "p50_latency": stats_entry.get_current_response_time_percentile(0.5),
+            "p90_latency": stats_entry.get_current_response_time_percentile(0.9),
+            "p99_latency": stats_entry.get_current_response_time_percentile(0.99),
+            "avg_rps": stats_entry.total_rps,
+        }
```
make this a dataclass please

edoakes commented on release/serve_tests/workloads/locust_utils.py:
```diff
@@ -0,0 +1,172 @@
+
+    def run(self):
+        import gevent
+        from locust.stats import (
+            get_stats_summary,
+            get_percentile_stats_summary,
+            get_error_report_summary,
+            stats_history,
+            stats_printer,
+        )
```
why lazy imports?

zcin replied:
For some reason if locust is imported in the driver script, connecting to ray will fail.

edoakes replied:
np leave a comment

edoakes commented on release/serve_tests/workloads/locust_utils.py:
```diff
@@ -0,0 +1,172 @@
+            )
+            time.sleep(1)
+
+        # Stats stuff
+        gevent.spawn(stats_printer(self.master_env.stats))
+        gevent.spawn(stats_history, master_runner)
```
what do we need gevent for? would prefer not to introduce another dependency here. should we just use asyncio?

zcin replied:
gevent is a locust dependency. The locust CLI uses gevent to print out stats every 5-10 seconds, so I mimicked it here so that we can get stats printed out in the job.

edoakes replied:
I see. seems fine. can you leave a better comment than "Stats stuff" :)

zcin commented:
@edoakes Addressed comments, please take another look when you get the chance!

edoakes commented on release/serve_tests/workloads/anyscale_service_utils.py:
```diff
@@ -0,0 +1,89 @@
+
+    if add_unique_suffix:
+        ray_version = (
+            ray.__commit__[:8]
+            if ray.__commit__ != "{{RAY_COMMIT_SHA}}"
+            else get_random_string()
```
probably should do both the SHA and random string, else multiple runs kicked off from the same commit have a race

zcin replied:
done

edoakes commented on release/serve_tests/workloads/anyscale_service_utils.py:
```diff
@@ -0,0 +1,89 @@
+
+    service_config = service.ServiceConfig(
+        name=service_name,
+        image_uri=f"anyscale/image/{cluster_env}:1",
```
can do `or None` here to default as i mentioned above

edoakes commented on release/serve_tests/workloads/anyscale_service_utils.py:
```diff
@@ -0,0 +1,89 @@
+    finally:
+        logger.info(f"Terminating service {service_name}.")
+        service.terminate(name=service_name)
+        wait_for_condition(
+            check_service_state,
+            service_name=service_name,
+            expected_state="TERMINATED",
+            retry_interval_ms=10000,  # 10s
+            timeout=600,
+        )
```
seems that probably we should add a util in the service SDK at some point to do this for folks

edoakes commented on release/serve_tests/workloads/autoscaling_load_test.py:
```diff
@@ -0,0 +1,102 @@
+        ],
+    }
+
+    with start_service(
+        "autoscaling-locust-test",
+        compute_config="cindy-all32-min0worker:1",
+        applications=[resnet_application],
+    ) as service_name:
```
very nice!

edoakes commented on release/serve_tests/workloads/autoscaling_load_test.py:
```diff
@@ -0,0 +1,102 @@
+
+    with start_service(
+        "autoscaling-locust-test",
+        compute_config="cindy-all32-min0worker:1",
```
we should inline the compute config when creating the service so everything is specified in one place. I've added a proper model for this in the SDK now. See `anyscale.compute_config.models.ComputeConfig`

zcin replied:
awesome!!

edoakes commented on release/serve_tests/workloads/autoscaling_load_test.py:
```diff
@@ -0,0 +1,102 @@
+            "total_requests": stats.total_requests,
+            "history": stats.history,
+            "perf_metrics": [
+                {
+                    "perf_metric_name": "avg_latency",
+                    "perf_metric_value": stats.avg_latency,
+                    "perf_metric_type": "LATENCY",
+                },
```
```suggestion
                {
                    "name": "avg_latency",
                    "value": stats.avg_latency,
                    "type": "LATENCY",
                },
```
nit: highly redundant

also -- should this be a dataclass too?

edoakes commented on release/serve_tests/workloads/autoscaling_load_test.py:
```diff
@@ -0,0 +1,102 @@
+
+if __name__ == "__main__":
+    main()
+    import pytest
+    import sys
```
nit: put at top of file

edoakes commented on release/serve_tests/workloads/autoscaling_load_test.py:
```diff
@@ -0,0 +1,102 @@
+    import pytest
+    import sys
+
+    sys.exit(pytest.main(["-v", "-s", __file__]))
```
why's there `main()` and this?

zcin replied:
Was unnecessary, removed code relating to pytest.

edoakes commented on release/serve_tests/workloads/locust_utils.py:
```diff
@@ -0,0 +1,263 @@
+
+        master_runner = self.master_env.create_master_runner("*", 5557)
+
+        while len(master_runner.clients.ready) < self.expected_num_workers:
+            print(
+                f"Waiting for workers to be ready, {len(master_runner.clients.ready)} "
+                f"of {self.expected_num_workers} ready."
+            )
+            time.sleep(1)
```
this should have some reasonable timeout and then error -- 10min? 15min?

edoakes replied:
probably make it configurable via env var or argument

zcin replied:
Made it configurable through the locust test config

edoakes commented on release/serve_tests/workloads/autoscaling_load_test.py:
```diff
@@ -0,0 +1,102 @@
+@click.command()
+def main():
+    resnet_application = {
+        "name": "default",
+        "import_path": "resnet_50:app",
+        "route_prefix": "/",
```
nit: drop unnecessary fields
```suggestion
        "import_path": "resnet_50:app",
```

edoakes commented on release/serve_tests/workloads/autoscaling_load_test.py:
```diff
@@ -0,0 +1,102 @@
+        "name": "default",
+        "import_path": "resnet_50:app",
+        "route_prefix": "/",
+        "runtime_env": {"working_dir": "workloads"},
```
nit: specify `working_dir` to `ServiceConfig` instead of in the runtime env. that way it can be shared for any application that gets passed in

edoakes reviewed (commented):
This is looking really great!!! All minor comments.

zcin commented on release/serve_tests/workloads/anyscale_service_utils.py:
```diff
@@ -0,0 +1,89 @@
+logger = logging.getLogger(__file__)
+logging.basicConfig(level=logging.INFO)
+
+DEFAULT_CLUSTER_ENV = "default_cluster_env_nightly-ml_py39"
```
Nice, thanks!

aslonnie commented on release/ray_release/job_manager/anyscale_job_manager.py:
```diff
@@ -55,6 +55,10 @@ def _run_job(
         upload_path: Optional[str] = None,
         pip: Optional[List[str]] = None,
     ) -> None:
+        env_vars[
+            "ANYSCALE_JOB_CLUSTER_ENV_NAME"
+        ] = self.cluster_manager.cluster_env_name
+
```
env_vars is a dict that is sent in, I do not think you are suppose to change its content.

could you make a copy of `env_vars` and then set the `ANYSCALE_JOB_CLUSTER_ENV_NAME`?

not sure, but maybe also add a unit test? (but if you cannot find it, it is fine.. @can-anyscale or I can add it later..)


also, I think the `env` related two lines can just be removed:

````
env = os.environ.copy()
env.setdefault("ANYSCALE_HOST", str(ANYSCALE_HOST))
````

those two lines do not seem to do anything useful...

zcin replied:
Sounds good, I've changed it to modify + pass in a copy of the env vars. I tried to add a unit test to `test_anyscale_job_manager.py` but couldn't find a way to execute it (kept getting `ModuleNotFoundError: No module named 'ray_release'` error)

zcin committed:
[serve] autoscaling release tests

Add autoscaling load test as serve release test. Starts a service and runs a locust workload against it.
Name: `serve_autoscaling_load_test`.

Signed-off-by: Cindy Zhang <cindyzyx9@gmail.com>

zcin commented:
@aslonnie Could you take another look when you get the chance? Thanks!

aslonnie commented:
<img width="236" alt="image" src="https://github.com/ray-project/ray/assets/95255098/d4e8e196-dbcd-4b99-a325-09f9f8132004">

just click the recycle button next time.

aslonnie reviewed (approved):
looks good for release test job manager change.

aslonnie commented:
> just click the recycle button next time.

I check github review requests more often than I check email..
